src/Raven.Server/Documents/Indexes/Static/IndexCompiler.cs

```diff
@@ -224,87 +226,203 @@ private static CompilationResult CompileInternal(string originalName, string cSh
+                newReferences.Add(References[i]);
+
+            for (int i = 0; i < knownManageRefs.Length; i++)
+                newReferences.Add(knownManageRefs[i].Reference);
+
+            if (definition.AdditionalAssemblies != null)
             {
-                newReferences[i] = References[i];
+                foreach (var additionalAssembly in definition.AdditionalAssemblies)
+                {
+                    if (additionalAssembly.AssemblyName != null)
+                    {
+                        newReferences.Add(FromAssemblyName(additionalAssembly.AssemblyName));
+                        continue;
+                    }
+
+                    if (additionalAssembly.AssemblyPath != null)
+                    {
+                        newReferences.Add(FromAssemblyPath(additionalAssembly.AssemblyPath));
+                        continue;
+                    }
+
+                    if (additionalAssembly.PackageName != null)
+                    {
+                        newReferences.AddRange(FromPackage(additionalAssembly.PackageName, additionalAssembly.PackageVersion, additionalAssembly.PackageSourceUrl));
+                        continue;
+                    }
+
+                    if (additionalAssembly.Usings != null && additionalAssembly.Usings.Count > 0)
+                        continue;
+
+                    throw new NotSupportedException($"Not supported additional assembly: {additionalAssembly}");
+                }
             }
-            for (int i = 0; i < knownManageRefs.Length; i++)
+
+            return newReferences.ToArray();
+
+            static MetadataReference FromAssemblyName(string name)
+            {
+                try
+                {
+                    var assemblyName = new AssemblyName(name);
+                    var assembly = Assembly.Load(assemblyName);
+                    return CreateMetadataReferenceFromAssembly(assembly);
+                }
+                catch (Exception e)
+                {
+                    throw new IndexCompilationException($"Cannot load assembly '{name}'.", e);
+                }
+            }
+
+            static MetadataReference FromAssemblyPath(string path)
             {
-                newReferences[i + References.Length] = knownManageRefs[i].Reference;
+                try
+                {
+                    var assembly = LoadAssembly(path);
+                    return CreateMetadataReferenceFromAssembly(assembly);
+                }
+                catch (Exception e)
+                {
+                    throw new IndexCompilationException($"Cannot load assembly from path '{path}'.", e);
+                }
+            }
+
+            static List<MetadataReference> FromPackage(string packageName, string packageVersion, string packageSourceUrl)
+            {
+                try
+                {
+                    packageSourceUrl ??= "https://api.nuget.org/v3/index.json";
```

Comment:
I think that this needs to be a configuration option.

Reply:
We can do that. Yes.